### PR TITLE
the delete key can delete 

### LINKED ISSUE #152: Add event listener for 'delete' key to delete item
Similar to 'enter' for navigation

### PR DESCRIPTION
**Pull request by Abdelra6**
fixes #152

### CODE REVIEW

**Comment by MahmoudHamdy02** [comment]
Hello @Abdelra6, thanks a lot for your contribution!

This needs a little more work:
1. You are calling `spawnDeleteItem` directly, instead of the wrapper function `deleteItem` in `fileActions.jsx`. You need to call `deleteItem` to show the confirmation dialog.
2. You don't need to delete every item separately in a for loop, `spawnDeleteItem` already supports deleting multiple items ([see here](https://github.com/cockpit-project/cockpit-navigator/blob/main/src/apis/spawnHelpers.jsx#L37))
3. Some general fixes which I will leave in a review

File: src/app.jsx [hunk 1]
```diff
@@ -17,6 +17,15 @@
  * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
  */
 
+import {
+    spawnCreateDirectory,
```
Comment (MahmoudHamdy02) [inline src/app.jsx:21]: Please only import the specific functions that you need

File: src/app.jsx [hunk 2]
```diff
@@ -447,7 +455,9 @@ const NavigatorCardBody = ({
     }
 
     const onDoubleClickNavigate = useCallback((file) => {
+        console.log("FFF");
```
Comment (MahmoudHamdy02) [inline src/app.jsx:458]: Global: Please remove all instances of `console.log`

File: src/app.jsx [hunk 3]
```diff
@@ -470,6 +480,7 @@ const NavigatorCardBody = ({
     });
 
     useEffect(() => {
+
```
Comment (MahmoudHamdy02) [inline src/app.jsx:483]: Global: Please remove extra empty lines

File: src/app.jsx [hunk 4]
```diff
@@ -508,8 +519,30 @@ const NavigatorCardBody = ({
                     return [sortedFiles[newIdx]];
                 });
             } else if (e.key === "Enter" && selected.length === 1) {
+                console.log("Entered");
                 onDoubleClickNavigate(selected[0]);
             }
+            else if (e.key === "Delete") {
+                const currentDir = path.join("/") + "/";
+                if(selected.length===1){    
+                const selectedContext=selected[0];
```
Comment (MahmoudHamdy02) [inline src/app.jsx:528]: Global: Please fix the formatting, linting errors should show up in the terminal where you run `make check`. 
I suggest you set up 'fix linting on save' in your editor to fix this automatically

**Review by MahmoudHamdy02 — CHANGES_REQUESTED** [review]

File: src/app.jsx (hunk 4 above)
Comment (Abdelra6) [reply src/app.jsx:528]: @MahmoudHamdy02 Thanks for the feedback I'll fix the formatting and about deleting multiple items I know it's not the most efficient and I know that it already supports it but I tried few things and wierd erros were popping after deleting multiple items

**Review by MahmoudHamdy02 — COMMENTED** [review]
Much better, thanks!

Pressing the `delete` key when no item is selected throws an error, the event should be ignored in this case. Also, there are still some `console.log`s and extra lines remaining, please remove.

**Comment by Abdelra6** [comment]
Sorry about the `console.log` . Removed them and handled the case when no of selected items is 0. Please tell me If there is anything else that needs fixing

File: src/app.jsx [hunk 5]
```diff
@@ -633,3 +652,4 @@ const NavigatorCardBody = ({
         );
     }
 };
```
Comment (KKoukiou) [inline src/app.jsx:654]: Unrelated change, please remove.

File: src/fileActions.jsx [hunk 6]
```diff
@@ -625,3 +625,4 @@ export const updateFile = (file, currentPath) => {
                     return file;
             }, exc => console.error(`Adding file ${file} failed: ${exc.toString()}`));
 };
```
Comment (KKoukiou) [inline src/fileActions.jsx:627]: Unrelated change, please remove.

File: src/app.jsx [hunk 7]
```diff
@@ -509,6 +509,22 @@ const NavigatorCardBody = ({
...
+                    const currentDir = path.join("/") + "/";
+                    deleteItem(
+                        Dialogs,
+                        {
+                            selected,
+                            itemPath: currentDir + selected[0].name,
+                            setHistory,
+                            setHistoryIndex,
+                            path: currentDir,
+                            multiple: Array.isArray(selected),
```
Comment (KKoukiou) [inline src/app.jsx:523]: Multiple seems to not be used

File: src/app.jsx [hunk 8]
```diff
@@ -509,6 +509,22 @@ const NavigatorCardBody = ({
...
             } else if (e.key === "Enter" && selected.length === 1) {
                 onDoubleClickNavigate(selected[0]);
+            } else if (e.key === "Delete") {
+                if (selected.length !== 0) {
+                    const currentDir = path.join("/") + "/";
+                    deleteItem(
+                        Dialogs,
+                        {
+                            selected,
+                            itemPath: currentDir + selected[0].name,
```
Comment (KKoukiou) [inline src/app.jsx:519]: We want to delete all selected items, not just the first one.

**Review by KKoukiou — CHANGES_REQUESTED** [review]

File: src/app.jsx (hunk 8 above)
Comment (Abdelra6) [reply src/app.jsx:519]: https://github.com/cockpit-project/cockpit-navigator/blob/e6145bee4e56709b94d8770dba7593ed7bf6f126/src/apis/spawnHelpers.jsx#L37-L39
Yup that was pretty confusing to me. But why does it work correctly (removes multiple items without error)? And I checked how "the regular delete" is being done and It seems to pass only one item so I did the same.

Comment (Abdelra6) [reply src/app.jsx:519]: I don't understand most of the codebase So please correct me If I'm misunderstanding something

File: src/fileActions.jsx (hunk 6 above)
Comment (Abdelra6) [reply src/fileActions.jsx:627]: what change? I didn't change anything in this file

Comment (KKoukiou) [reply src/fileActions.jsx:627]: I see a new line was added. Probbaly your editor did it.

Comment (Abdelra6) [reply src/fileActions.jsx:627]: Ah! got it thanks

Comment (Abdelra6) [reply src/fileActions.jsx:627]: I get an eslint error when I remove that last line. So I thought it should be this way

File: src/app.jsx (hunk 8 above)
Comment (MahmoudHamdy02) [reply src/app.jsx]: @KKoukiou `itemPath` is ignored if multiple items are selected, so it's okay to pass `selected[0].name`. `spawnDeleteItem` handles multiple items.
The previous commit is works correctly, we can drop the latest commit. Let me know what you think!

`mutliple` is unused indeed, it's also present in the [main delete button,](https://github.com/cockpit-project/cockpit-navigator/blob/main/src/app.jsx#L215) we can remove it here or in a follow up.

**Comment by Abdelra6** [comment]
actually it shows the confirmation dialoug in the last commit. That's a video of it If I understand what you mean correctly
[Screencast from 2024-01-22 20-05-26.webm](https://github.com/cockpit-project/cockpit-navigator/assets/43535792/d95168d7-7a57-4aef-b760-77d0c0194d87)

**Comment by jelly** [comment]
> actually it shows the confirmation dialoug in the last commit. That's a video of it If I understand what you mean correctly [Screencast from 2024-01-22 20-05-26.webm](https://github.com/cockpit-project/cockpit-navigator/assets/43535792/d95168d7-7a57-4aef-b760-77d0c0194d87)

Aha, well I only selected one file so I guess this is a bug? It would be great if you could rebase the PR against main.

**Comment by Abdelra6** [comment]
It also works if you selected single files. Maybe you fetched some older commit or smth. Because that's where the confirmation dialog was not showing. I also rebased it to main)

**Comment by jelly** [comment]
> It also works if you selected single files. Maybe you fetched some older commit or smth. Because that's where the confirmation dialog was not showing. I also rebased it to main)

Re-tested and indeed it works, for this to land it would be nice to have some automated tests as well but sending a delete key seems to be non-trivial with our test framework.

File: src/app.jsx [hunk 9]
```diff
@@ -494,6 +494,35 @@ const NavigatorCardBody = ({
...
+                        deleteItem(
+                            Dialogs,
+                            {
+                                selected,
+                                itemPath: currentDir + selected.map(s => currentDir + s.name),
+                                setHistory,
+                                setHistoryIndex,
+                                path: currentDir,
+                                setSelected
+                            }
```
Comment (jelly) [inline src/app.jsx:522]: Both conditionals are almost the same, and can be combined into one as selected is always an array.

File: src/app.jsx [hunk 10]
```diff
@@ -494,6 +494,35 @@ const NavigatorCardBody = ({
...
+                        );
+                    } else {
+                        deleteItem(
+                            Dialogs,
+                            {
+                                selected,
+                                itemPath: currentDir + selected.map(s => currentDir + s.name),
```
Comment (jelly) [inline src/app.jsx:517]: This doesn't appear to make sense:
```
currentDir + selected.map(s => currentDir + s.name)
'/home/admin/man7//home/admin/man7/EVP_KEYMGMT-Poly1305.7ossl.gz,/home/admin/man7/EVP_KEYMGMT-RSA.7ossl.gz'
```

The code in sidebar.jsx passes `selected` if multiple files are selected.

Comment (jelly) [reply src/app.jsx:517]: Take a look at the existing `deleteItem` invocation in `src/sidebar.jsx` for [one item](https://github.com/cockpit-project/cockpit-navigator/blob/87e2fc5ca94e0234689e1a485044873afeedfb13/src/sidebar.jsx#L342) and for [multiple](https://github.com/cockpit-project/cockpit-navigator/blob/87e2fc5ca94e0234689e1a485044873afeedfb13/src/sidebar.jsx#L368).

So basically both can be combined and the only difference is that for `selected.length === 1` we need to provide the full path and otherwise itemPath can be left out.

Comment (Abdelra6) [reply src/app.jsx]: I made a last commit. Could I just leave it and work on another issue?

Comment (jelly) [reply src/app.jsx]: There is still an unresolved issue in the PR about combing the if/else into one deleteItem

Comment (Abdelra6) [reply src/app.jsx]: > Actually that's what I was doing exactly in the commit before the last one .but KKoukiou told me that what if multiple files are selected that's why I changed it.so I think we can Ignore the last commit as Mahmoud said above or I still don't know how things work

that's why I replied with this. Anyways I'm going to make another commit

**Comment by Abdelra6** [comment]
Is this one done?

**Comment by subhoghoshX** [comment]
@Abdelra6 continuing our conversation from linkedin/matrix, I've rebased the PR and wrote automated tests. It's working with chromium but not in firefox (the reason testing-farm failling) :(

I did some digging, [following pitti's blog](https://piware.de/post/2017-12-21-phantomjs-to-chromium/), and found some inconsistencies between the browsers.

In case of chromium the method `Input.dispatchKeyEvent({type: 'keyDown', key: '\x7f'})` fires the event with `event.key === 'Delete'` but in case of firefox `event.key` is empty string (or maybe some unknown character)..

The only way I've found they're consistent is if we pass "Delete" explicitly like `Input.dispatchKeyEvent({type: 'keyDown', key: 'Delete'})` instead of the ascii equivalent '\x7f' (which is the return value of python `chr(127)`). But `b.key_press('Delete')` wouldn't work because our test framework [loops trough the string](https://github.com/cockpit-project/cockpit/blob/main/test/common/testlib.py#L502) and sends each character separately so it'd become `D` `e` `l` `e` `t` `e`.

It's trivial to fix with introducing another parameter or some conditionals in `def key_press()` but I was trying to avoid touching that file.

@jelly any thoughts?

**Comment by jelly** [comment]
> @Abdelra6 continuing our conversation from linkedin/matrix, I've rebased the PR and wrote automated tests. It's working with chromium but not in firefox (the reason testing-farm failling) :(
> 
> I did some digging, [following pitti's blog](https://piware.de/post/2017-12-21-phantomjs-to-chromium/), and found some inconsistencies between the browsers.
> 
> In case of chromium the method `Input.dispatchKeyEvent({type: 'keyDown', key: '\x7f'})` fires the event with `event.key === 'Delete'` but in case of firefox `event.key` is empty string (or maybe some unknown character)..
> 
> The only way I've found they're consistent is if we pass "Delete" explicitly like `Input.dispatchKeyEvent({type: 'keyDown', key: 'Delete'})` instead of the ascii equivalent '\x7f' (which is the return value of python `chr(127)`). But `b.key_press('Delete')` wouldn't work because our test framework [loops trough the string](https://github.com/cockpit-project/cockpit/blob/main/test/common/testlib.py#L502) and sends each character separately so it'd become `D` `e` `l` `e` `t` `e`.
> 
> It's trivial to fix with introducing another parameter or some conditionals in `def key_press()` but I was trying to avoid touching that file.
> 
> @jelly any thoughts?

I am happy with just calling `self.cdp.invoke("Input.dispatchKeyEvent", **args)` here. Ideally we would be able to press Delete and also ArrowUp/Down with `key_press` but I've also found some issues testing these keys using `testlib`. So maybe once we figure out multiple special keys we can either introduce press_delete() or fix `key_press`.

**Comment by jelly** [comment]
Fixed via https://github.com/cockpit-project/cockpit-files/pull/358
